ipq40xx: add support for MobiPromo CM520-79F

dengqf6 opened the pull request:
MobiPromo CM520-79F is an AC1200 dual band router based on IPQ4019

Specification:

SoC/Wireless: QCA IPQ4019
RAM: 512MiB
Flash: 128MiB SLC NAND
Ethernet PHY: QCA8075
Ethernet ports: 1x WAN, 2x LAN
LEDs: 7 LEDs
      2 (USB, CAN) are GPIO
      other 5 (2.4G, 5G, LAN1, LAN2, WAN) are connected to a shift register
Button: Reset

Flash instruction:
Disassemble the router, connect UART pins like this:
```
 GND TX    RX
  [x x . . x .]
  [. . . . . .]

(QCA8075 and IPQ4019 below)
```
Baud-rate: 115200 8-N-1

Set up TFTP server: IP 192.168.1.183/24
Power on the router and interrupt the booting with UART console
Boot initramfs from TFTP:
```
tftpboot openwrt-ipq40xx-generic-mobipromo_cm520-79f-initramfs-fit-zImage.itb
bootm
```
After initramfs image is booted, backup rootfs partition in case of reverting to stock image
```
cat /dev/mtd12 > /tmp/mtd12.bin
cat /dev/mtd13 > /tmp/mtd13.bin
```
Then fetch it via SCP

Upload nand-factory.ubi to /tmp via SCP, then run
```
mtd erase rootfs
mtd write /tmp/*nand-factory.ubi rootfs
reboot
```
To revert to stock image, upload the backed up bin and run
```
mtd write /tmp/mtd12.bin rootfs
mtd write /tmp/mtd13.bin rootfs_1
reboot
```
The BOOTCONFIG may have been configured to boot from alternate partition (rootfs_1) instead
In case of this, set it back to rootfs:
```
cd /tmp
cat /dev/mtd7 > mtd7.bin
echo -ne '\x0b' | dd of=mtd7.bin conv=notrunc bs=1 count=1 seek=4
for i in 28 48 68 108
	do dd if=/dev/zero of=mtd7.bin conv=notrunc bs=1 count=1 seek=$i
done
mtd write mtd7.bin BOOTCONFIG
mtd write mtd7.bin BOOTCONFIG1
```

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -55,6 +55,11 @@ ipq40xx_setup_interfaces()
+	mobipromo,cm520-79f)
+		ucidef_set_interfaces_lan_wan "eth0" "eth1"
+		ucidef_add_switch "switch0" \
+			"0u@eth0" "3:lan:2" "4:lan:1"
```
As far as I know, ucidef_add_switch will already set up "lan" here, it should be enough to set "wan" explicitly before:
ucidef_set_interfaces_wan "eth1"

dengqf6 replied:
Done

adrianschmutzler commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-cm520-79f.dts:
```diff
@@ -0,0 +1,365 @@
+};
+
+&gmac0 {
+	mtd-mac-address = <&art 0x1006>;
```
So, the two WiFi addresses are used for the two GMACs? Interesting ...

Does the device have a label/sticker/box with MAC address on it? If yes, please tell which it is.

dengqf6 replied:
No.. And its crappy stock firmware uses a random one..

adrianschmutzler replied:
Ah, so _you_ "decided" about the MAC addresses for &gmac0 and &gmac1 ...
Well, better than a random one.

chunkeey replied:
heh, I think this will need to be redone once nvmem DT gets supported.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -55,6 +55,11 @@ ipq40xx_setup_interfaces()
+	mobipromo,cm520-79f)
+		ucidef_add_switch "switch0" \
+			"0u@eth0" "3:lan:2" "4:lan:1"
+		ucidef_set_interface_wan "eth1"
```
Typically, ucidef_set_interface_wan is put before ucidef_add_switch. But that's a cosmetical issue.

dengqf6 replied:
I put `ucidef_set_interface_wan` after `ucidef_add_switch` so that interface wan is added after lan.

adrianschmutzler commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-cm520-79f.dts:
```diff
@@ -0,0 +1,360 @@
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-cm520-79f.dts:
```diff
@@ -0,0 +1,363 @@
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+
```
Remove empty line.

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -26,6 +26,7 @@ endef
 ALLWIFIBOARDS:= \
 	avm_fritzrepeater-1200 \
 	linksys_ea8300 \
+	mobipromo_cm520-79f \
```
Please follow the instructions above in the comment and post the vendors boardfiles to the ath10k mailing list

>
> This is intended to be used on an interim basis until device-specific
> board data for new devices is available through the upstream compilation
>
> Please send a mail with your device-specific board files upstream.
> You can find instructions and examples on the linux-wireless wiki:
> <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
>

swg0101 replied:
Board files are now sent to the ath10k mailing list.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-cm520-79f.dts:
```diff
@@ -0,0 +1,354 @@
+		compatible = "spi-gpio";
+		#address-cells = <1>;
+
+		gpio-sck = <&tlmm 40 GPIO_ACTIVE_HIGH>;
```
Oh boy, could you please mention that this will need to be updated because the binding will change?

https://www.kernel.org/doc/Documentation/devicetree/bindings/spi/spi-gpio.txt

gpio-sck -> sck-gpios
gpio-mosi -> mosi->gpios

adrianschmutzler replied:
>  But I don't think we need this file anymore, since new devices don't get backported?!

@chunkeey AFAIK there is no clear recommendation on this topic, and frequently both files are submitted. I personally do not see much use in having these files twice, as I don't think building master with 4.14 is a common scenarios. It will just make reviewing more complicated. 
So, IMO it would be helpful to somewhat agree on only requiring 4.19 files here. I'm also not really sure what's the point of still keeping the 4.14 files around, as backporting changes shouldn't be that hard without them.

chunkeey replied:
> I'm also not really sure what's the point of still keeping the 4.14 files around, as backporting changes shouldn't be that hard without them.

Oh, from what I remember the decision to keep the old kernel around came about because of a complaint because the stable release kernel broke and the fix was non-trivial (can't find the post on the ML sadly - I think it was on the openwrt-adm though). So the kernel patches (and support) of the target should be kept around to make it possible to switch to the currenty/next stable release (19.07) for testing/developing to help out the backporters.

adrianschmutzler replied:
Mhm, but this would not restrict us from at least having new support 4.19 only (or whatever up-to-date, unless TESTING_PATCHVER).

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-cm520-79f.dts:
```diff
@@ -0,0 +1,354 @@
+
+	led_spi {
+		compatible = "spi-gpio";
+		#address-cells = <1>;
```
you need the "ranges;" property as well.

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -324,6 +324,18 @@ define Device/meraki_mr33
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := MobiPromo
+	DEVICE_MODEL := CM520-79F
+	DEVICE_DTS := qcom-ipq4019-cm520-79f
```
Replace this with 

SOC := qcom-ipq4019

dengqf6 commented:
Rebased and removed 4.14 files.

Commit:
ipq40xx: add support for MobiPromo CM520-79F

MobiPromo CM520-79F is an AC1300 dual band router based on IPQ4019

Specification:

SoC/Wireless: QCA IPQ4019
RAM: 512MiB
Flash: 128MiB SLC NAND
Ethernet PHY: QCA8075
Ethernet ports: 1x WAN, 2x LAN
LEDs: 7 LEDs
      2 (USB, CAN) are GPIO
      other 5 (2.4G, 5G, LAN1, LAN2, WAN) are connected to a shift register
Button: Reset

Flash instruction:
Disassemble the router, connect UART pins like this:
 GND TX    RX
  [x x . . x .]
  [. . . . . .]

(QCA8075 and IPQ4019 below)
Baud-rate: 115200

Set up TFTP server: IP 192.168.1.188/24
Power on the router and interrupt the booting with UART console
Set bootenv:
	setenv set_ubi 'set mtdids nand0=nand0; set mtdparts mtdparts=nand0:0x7480000@0xb80000(fs); ubi part fs'
	setenv bootkernel 'ubi read 0x84000000 kernel; bootm 0x84000000#config@1'
	setenv bootargs 'ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs'
	setenv cm520_boot 'run set_ubi; run bootkernel'
	setenv bootcmd 'run cm520_boot'
	saveenv
Boot initramfs from TFTP:
	tftpboot openwrt-ipq40xx-generic-mobipromo_cm520-79f-initramfs-fit-zImage.itb
	bootm
After initramfs image is booted, backup rootfs partition in case of reverting to stock image
	cat /dev/mtd12 > /tmp/mtd12.bin
Then fetch it via SCP

Upload nand-factory.ubi to /tmp via SCP, then run
	mtd erase rootfs
	mtd write /tmp/*nand-factory.ubi rootfs
	reboot

To revert to stock image, restore default bootenv in uboot UART console
	setenv bootcmd 'bootipq'
	saveenv
upload the backed up mtd12.bin and run
	tftpboot mtd12.bin
	nand erase 0xb80000 0x7480000
	nand write 0x84000000 0xb80000 0x7480000
The BOOTCONFIG may have been configured to boot from alternate partition (rootfs_1) instead
In case of this, set it back to rootfs:
	cd /tmp
	cat /dev/mtd7 > mtd7.bin
	echo -ne '\x0b' | dd of=mtd7.bin conv=notrunc bs=1 count=1 seek=4
	for i in 28 48 68 108; do
		dd if=/dev/zero of=mtd7.bin conv=notrunc bs=1 count=1 seek=$i
	done
	mtd write mtd7.bin BOOTCONFIG
	mtd write mtd7.bin BOOTCONFIG1

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>

chunkeey commented:
Did you sent the boardfile to the ath10k mailing list yet?

swg0101 commented:
@dengqf6 Can you please rebase again?

swg0101 commented:
@chunkeey Is the only thing left to send the board file to the mailing list?

chunkeey commented:
> @chunkeey Is the only thing left to send the board file to the mailing list?

To the ath10k mailing list. For more information about the procedure, see the [boardfile](https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles) wiki article over at the official linux-wireless wiki. 

As for the device, there's one other thing as well that I'm not really sure. So this is more of a suggestion to test something.

The DTS contains:

    bootargs-append = " ubi.block=0,1 root=/dev/ubiblock0_1";

Now, this bootargs property might be avoidable. Since the "rootfs" partition could be renamed to "ubi" which would make it possible for the kernel to autoattach and automount the ubi volumes (because OpenWrt has a special patch for this). But the author didn't do this, so I don't know if there's something special about this device or if it was just overlooked.

swg0101 commented:
@chunkeey I should have the hardware in about a week. At that time I should be able to verify the board files and the bootargs. Thanks!

swg0101 commented:
@chunkeey I have just sent the board files over onto the ath10k mailing lists. I have also confirmed that the ones attached to this PR commit are identical to the ones present in the stock firmware.
496396e221f0c0550899fa851e507b92ffdfb8d671ff7a6734bedd31b96e862e    /lib/firmware/IPQ4019/hw.1/boardData_1_0_IPQ4019_DK04_2G.bin - bus=ahb,bmi-chip-id=0,bmi-board-id=20,variant=CM520-79F.bin
c9a0bf20c92ac3891bbb3a3b6a9500d95dd678c98b7118a65b568d3bb06adb6d    /lib/firmware/IPQ4019/hw.1/boardData_1_0_IPQ4019_DK04_5G.bin - bus=ahb,bmi-chip-id=0,bmi-board-id=21,variant=CM520-79F.bin

As for your other concern, I did confirm that the extra bootargs were not required when the rootfs was renamed to the "ubi" label. Although I can't speak for the author himself, I believe the reason why he chose to have this added was because the uboot env has this specified from stock:
bootargs=ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs rootwait
and that the author wanted to keep to that naming convention (in case someone wants to restore back to stock). Stock is actually using Chaos Calmer as a base, which is a bit old. To get the new modified build (with the ubi label) to boot, I had to deviate from the commit instructions of unsetting the bootargs:
setenv bootargs
saveenv

It is possible that he was using that as a template since many other devices seem to be booting using a similar method.

It also seems like that the author has modified the DTS so that the rootfs (ubi) now takes up the rest of the NAND instead of having two distinct root partitions (which probably doesn't make a lot of sense for the official OpenWrt build which doesn't switch between A & B automatically).

@dengqf6 - Can you comment on this?

chunkeey commented:
Thanks, I've put it with the aforementioned changes into my tree (the one that goes upstream).
If I don't hear anything, I'll push it on the weekend when I can watch the builders.

swg0101 commented:
Thanks @chunkeey - I have been building this on the side over the last few weeks and the builds appear to be stable. The router will reboot if I subject it to arbitrarily low voltages, but that's not an OpenWrt issue... ;)

chunkeey commented:
Thanks, I haven't heard of any problems with the modification. So, I've merged the patch. If something comes up, please let me know.

ldhtrf commented:
Thanks to everyone who are involved with this pull request, but unfortunately I can't get the image to boot on my device after writing the ubi image to mtd12:

```
[    1.379409] mtd:ubi_rootfs: Can't open blockdeverror -2
[    1.379445] VFS: Cannot open root device "mtd:ubi_rootfs" or unknown-block(0,0): error -2
[    1.382748] Please append a correct "root=" boot option; here are the available partitions:
[    1.391123] 1f00            1024 mtdblock0 
[    1.391126]  (driver?)
[    1.403405] 1f01            1024 mtdblock1 
[    1.403408]  (driver?)
[    1.409915] 1f02            1024 mtdblock2 
[    1.409917]  (driver?)
[    1.416439] 1f03            1024 mtdblock3 
[    1.416441]  (driver?)
[    1.422935] 1f04            1024 mtdblock4 
[    1.422937]  (driver?)
[    1.429447] 1f05             512 mtdblock5 
[    1.429449]  (driver?)
[    1.435972] 1f06             512 mtdblock6 
[    1.435975]  (driver?)
[    1.442466] 1f07             512 mtdblock7 
[    1.442468]  (driver?)
[    1.448976] 1f08             512 mtdblock8 
[    1.448978]  (driver?)
[    1.455498] 1f09            2048 mtdblock9 
[    1.455501]  (driver?)
[    1.461997] 1f0a            2048 mtdblock10 
[    1.461999]  (driver?)
[    1.468866] 1f0b             512 mtdblock11 
[    1.468868]  (driver?)
[    1.475463] 1f0c          119296 mtdblock12 
[    1.475465]  (driver?)
[    1.482049] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
```

I have tried modifying and unsetting `bootargs` and `fsbootargs` in uboot, but they didn't seem to have any effect on this error: `mtd:ubi_rootfs: Can't open blockdeverror -2`. It always says `mtd:ubi_rootfs` no matter what I do. The image I used was the latest snapshot build (sha256 = 7160e3d1e53938bfcb791518d49fb1f176ccbc63e399b833a6e304d96373d3d8).

swg0101 commented:
The commit uses @dengqf6's commit instructions, which states:

```
Set bootenv:
	setenv set_ubi 'set mtdids nand0=nand0; set mtdparts mtdparts=nand0:0x7480000@0xb80000(fs); ubi part fs'
	setenv bootkernel 'ubi read 0x84000000 kernel; bootm 0x84000000#config@1'
	setenv bootargs 'ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs'
	setenv cm520_boot 'run set_ubi; run bootkernel'
	setenv bootcmd 'run cm520_boot'
	saveenv
```

Are you sure you added these correctly?

ldhtrf commented:
I tried setting the variables again but it still did not work. This is what my env looks like:
```
(IPQ40xx) # printenv
baudrate=115200
bootargs=ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs
bootcmd=run cm520_boot
bootdelay=2
bootkernel=ubi read 0x84000000 kernel; bootm 0x84000000#config@1
cm520_boot=run set_ubi; run bootkernel
ethact=eth0
fdt_high=0x87000000
fileaddr=88000000
filesize=5C7DF
flash_type=2
fsbootargs=ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs
ipaddr=192.168.1.183
machid=8010001
mtddevname=fs
mtddevnum=0
mtdids=nand0=nand0
mtdparts=mtdparts=nand0:0x4100000@0xb80000(fs),
partition=nand0,0
serverip=192.168.1.10
set_ubi=set mtdids nand0=nand0; set mtdparts mtdparts=nand0:0x7480000@0xb80000(fs); ubi part fs
stderr=serial
stdin=serial
stdout=serial

Environment size: 718/262140 bytes
```

swg0101 commented:
Can you post the boot log again from that?

ldhtrf commented:
Sure, here it is:

```
Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 5.4.31 (builder@buildhost) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r12968-a5a90a94ce)) #0 SMP Mon Apr 13 23:55:02 2020
[    0.000000] CPU: ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
[    0.000000] CPU: div instructions available: patching division code
[    0.000000] CPU: PIPT / VIPT nonaliasing data cache, VIPT aliasing instruction cache
[    0.000000] OF: fdt: Machine model: MobiPromo CM520-79F
[    0.000000] Memory policy: Data cache writealloc
[    0.000000] percpu: Embedded 15 pages/cpu s30028 r8192 d23220 u61440
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129408
[    0.000000] Kernel command line: ubi.mtd=rootfs root=mtd:ubi_rootfs rootfstype=squashfs
[    0.000000] Dentry cache hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 507248K/522240K available (5693K kernel code, 185K rwdata, 1500K rodata, 1024K init, 238K bss, 14992K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 16, nr_irqs: 16, preallocated irqs: 16
[    0.000000] random: get_random_bytes called from start_kernel+0x2f0/0x49c with crng_init=0
[    0.000000] arch_timer: cp15 timer(s) running at 48.00MHz (virt).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0xb11fd3bfb, max_idle_ns: 440795203732 ns
[    0.000009] sched_clock: 56 bits at 48MHz, resolution 20ns, wraps every 4398046511096ns
[    0.000023] Switching to timer-based delay loop, resolution 20ns
[    0.000243] Calibrating delay loop (skipped), value calculated using timer frequency.. 96.00 BogoMIPS (lpj=480000)
[    0.000259] pid_max: default: 32768 minimum: 301
[    0.000466] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.000482] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.001488] CPU: Testing write buffer coherency: ok
[    0.002484] Setting up static identity map for 0x80300000 - 0x80300060
[    0.002617] rcu: Hierarchical SRCU implementation.
[    0.003045] smp: Bringing up secondary CPUs ...
[    0.006067] smp: Brought up 1 node, 4 CPUs
[    0.006085] SMP: Total of 4 processors activated (384.00 BogoMIPS).
[    0.006093] CPU: All CPU(s) started in SVC mode.
[    0.010754] VFP support v0.3: implementor 41 architecture 2 part 30 variant 7 rev 5
[    0.010907] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.010933] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.011203] pinctrl core: initialized pinctrl subsystem
[    0.012274] NET: Registered protocol family 16
[    0.012583] DMA: preallocated 256 KiB pool for atomic coherent allocations
[    0.013779] cpuidle: using governor ladder
[    0.013823] cpuidle: using governor menu
[    0.035603] cryptd: max_cpu_qlen set to 1000
[    0.042500] usbcore: registered new interface driver usbfs
[    0.042573] usbcore: registered new interface driver hub
[    0.042659] usbcore: registered new device driver usb
[    0.042709] pps_core: LinuxPPS API ver. 1 registered
[    0.042719] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.042746] PTP clock support registered
[    0.044337] clocksource: Switched to clocksource arch_sys_counter
[    0.045130] thermal_sys: Registered thermal governor 'step_wise'
[    0.045455] NET: Registered protocol family 2
[    0.046254] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.046302] TCP established hash table entries: 4096 (order: 2, 16384 bytes, linear)
[    0.046365] TCP bind hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.046453] TCP: Hash tables configured (established 4096 bind 4096)
[    0.046584] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.046623] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.046859] NET: Registered protocol family 1
[    0.046901] PCI: CLS 0 bytes, default 64
[    0.049595] workingset: timestamp_bits=14 max_order=17 bucket_order=3
[    0.055360] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.055376] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.107347] bam-dma-engine 8e04000.dma: num-channels unspecified in dt
[    0.107367] bam-dma-engine 8e04000.dma: num-ees unspecified in dt
[    0.108632] tcsr 1949000.tcsr: setting wifi_glb_cfg = 41000000
[    0.108732] tcsr 194b000.tcsr: setting usb hs phy mode select = e700e7
[    0.108816] tcsr 1953000.ess_tcsr: setting ess interface select = 0
[    0.108896] tcsr 1957000.tcsr: setting wifi_noc_memtype_m0_m2 = 2222222
[    0.109218] Serial: 8250/16550 driver, 16 ports, IRQ sharing enabled
[    0.110988] msm_serial 78af000.serial: msm_serial: detected port #0
[    0.111031] msm_serial 78af000.serial: uartclk = 1843200
[    0.111089] 78af000.serial: ttyMSM0 at MMIO 0x78af000 (irq = 24, base_baud = 115200) is a MSM
[    0.111118] msm_serial: console setup on port #0
[    0.584330] printk: console [ttyMSM0] enabled
[    0.589214] msm_serial 78b0000.serial: msm_serial: detected port #1
[    0.593261] msm_serial 78b0000.serial: uartclk = 1843200
[    0.599384] 78b0000.serial: ttyMSM1 at MMIO 0x78b0000 (irq = 25, base_baud = 115200) is a MSM
[    0.605162] msm_serial: driver initialized
[    0.618268] loop: module loaded
[    0.619846] nand: device found, Manufacturer ID: 0x2c, Chip ID: 0xf1
[    0.620318] nand: Micron MT29F1G08ABAEAWP
[    0.626952] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    0.630926] 13 fixed-partitions partitions found on MTD device qcom_nand.0
[    0.638317] Creating 13 MTD partitions on "qcom_nand.0":
[    0.645165] 0x000000000000-0x000000100000 : "SBL1"
[    0.652454] 0x000000100000-0x000000200000 : "MIBIB"
[    0.656996] 0x000000200000-0x000000300000 : "BOOTCONFIG"
[    0.661734] 0x000000300000-0x000000400000 : "QSEE"
[    0.667308] 0x000000400000-0x000000500000 : "QSEE_1"
[    0.671853] 0x000000500000-0x000000580000 : "CDT"
[    0.676610] 0x000000580000-0x000000600000 : "CDT_1"
[    0.681172] 0x000000600000-0x000000680000 : "BOOTCONFIG1"
[    0.685924] 0x000000680000-0x000000700000 : "APPSBLENV"
[    0.691417] 0x000000700000-0x000000900000 : "APPSBL"
[    0.697808] 0x000000900000-0x000000b00000 : "APPSBL_1"
[    0.702974] 0x000000b00000-0x000000b80000 : "ART"
[    0.706656] 0x000000b80000-0x000008000000 : "ubi"
[    0.713724] random: fast init done
[    0.817694] libphy: ipq40xx_mdio: probed
[    0.848784] ESS reset ok!
[    0.881780] ESS reset ok!
[    1.299775] libphy: Fixed MDIO Bus: probed
[    1.304864] i2c /dev entries driver
[    1.306920] sdhci: Secure Digital Host Controller Interface driver
[    1.307161] sdhci: Copyright(c) Pierre Ossman
[    1.313408] sdhci-pltfm: SDHCI platform and OF driver helper
[    1.323115] NET: Registered protocol family 10
[    1.324956] Segment Routing with IPv6
[    1.327901] NET: Registered protocol family 17
[    1.331602] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    1.336107] 8021q: 802.1Q VLAN Support v1.8
[    1.349108] Registering SWP/SWPB emulation handler
[    1.364025] UBI error: cannot open mtd rootfs, error -2[    1.369306] mtd:ubi_rootfs: Can't open blockdev
[    1.369342] VFS: Cannot open root device "mtd:ubi_rootfs" or unknown-block(0,0): error -2
[    1.372642] Please append a correct "root=" boot option; here are the available partitions:
[    1.381022] 1f00            1024 mtdblock0 
[    1.381025]  (driver?)
[    1.393303] 1f01            1024 mtdblock1 
[    1.393306]  (driver?)
[    1.399810] 1f02            1024 mtdblock2 
[    1.399812]  (driver?)
[    1.406336] 1f03            1024 mtdblock3 
[    1.406338]  (driver?)
[    1.412831] 1f04            1024 mtdblock4 
[    1.412834]  (driver?)
[    1.419342] 1f05             512 mtdblock5 
[    1.419344]  (driver?)
[    1.425868] 1f06             512 mtdblock6 
[    1.425870]  (driver?)
[    1.432362] 1f07             512 mtdblock7 
[    1.432364]  (driver?)
[    1.438874] 1f08             512 mtdblock8 
[    1.438876]  (driver?)
[    1.445393] 1f09            2048 mtdblock9 
[    1.445395]  (driver?)
[    1.451893] 1f0a            2048 mtdblock10 
[    1.451895]  (driver?)
[    1.458752] 1f0b             512 mtdblock11 
[    1.458755]  (driver?)
[    1.465359] 1f0c          119296 mtdblock12 
[    1.465361]  (driver?)
[    1.471944] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    1.474130] CPU1: stopping
[    1.482365] CPU: 1 PID: 0 Comm: swapper/1 Not tainted 5.4.31 #0
[    1.485053] Hardware name: Generic DT based system
[    1.490892] [<c030e524>] (unwind_backtrace) from [<c030afa8>] (show_stack+0x10/0x14)
[    1.495745] [<c030afa8>] (show_stack) from [<c0873dd8>] (dump_stack+0x94/0xa8)
[    1.503642] [<c0873dd8>] (dump_stack) from [<c030d7d8>] (handle_IPI+0x164/0x19c)
[    1.510676] [<c030d7d8>] (handle_IPI) from [<c05a8d4c>] (gic_handle_irq+0x8c/0x90)
[    1.518226] [<c05a8d4c>] (gic_handle_irq) from [<c0301a8c>] (__irq_svc+0x6c/0x90)
[    1.525596] Exception stack(0xdf471f60 to 0xdf471fa8)
[    1.533152] 1f60: 00000000 000001a0 dfb461f4 c0313be0 ffffe000 c0c04e68 c0c04ea8 00000002
[    1.538190] 1f80: 00000000 c0b3ede8 c0c0dcf4 00000000 1f006000 df471fb0 c03083e8 c03083ec
[    1.546342] 1fa0: 60000013 ffffffff
[    1.554506] [<c0301a8c>] (__irq_svc) from [<c03083ec>] (arch_cpu_idle+0x38/0x3c)
[    1.557811] [<c03083ec>] (arch_cpu_idle) from [<c03495b0>] (do_idle+0x1e4/0x21c)
[    1.565446] [<c03495b0>] (do_idle) from [<c0349878>] (cpu_startup_entry+0x18/0x1c)
[    1.572822] [<c0349878>] (cpu_startup_entry) from [<803024ac>] (0x803024ac)
[    1.580199] CPU0: stopping
[    1.587050] CPU: 0 PID: 0 Comm: swapper/0 Not tainted 5.4.31 #0
[    1.589827] Hardware name: Generic DT based system
[    1.595654] [<c030e524>] (unwind_backtrace) from [<c030afa8>] (show_stack+0x10/0x14)
[    1.600515] [<c030afa8>] (show_stack) from [<c0873dd8>] (dump_stack+0x94/0xa8)
[    1.608414] [<c0873dd8>] (dump_stack) from [<c030d7d8>] (handle_IPI+0x164/0x19c)
[    1.615444] [<c030d7d8>] (handle_IPI) from [<c05a8d4c>] (gic_handle_irq+0x8c/0x90)
[    1.623000] [<c05a8d4c>] (gic_handle_irq) from [<c0301a8c>] (__irq_svc+0x6c/0x90)
[    1.630368] Exception stack(0xc0c01f28 to 0xc0c01f70)
[    1.637922] 1f20:                   00000000 00002638 dfb371f4 c0313be0 ffffe000 c0c04e68
[    1.642967] 1f40: c0c04ea8 00000001 00000000 c0b3ede8 c0c0dcf4 00000000 c0c04f14 c0c01f78
[    1.651119] 1f60: c03083e8 c03083ec 60000013 ffffffff
[    1.659280] [<c0301a8c>] (__irq_svc) from [<c03083ec>] (arch_cpu_idle+0x38/0x3c)
[    1.664318] [<c03083ec>] (arch_cpu_idle) from [<c03495b0>] (do_idle+0x1e4/0x21c)
[    1.671784] [<c03495b0>] (do_idle) from [<c0349878>] (cpu_startup_entry+0x18/0x1c)
[    1.679166] [<c0349878>] (cpu_startup_entry) from [<c0b00dfc>] (start_kernel+0x490/0x49c)
[    1.686542] [<c0b00dfc>] (start_kernel) from [<00000000>] (0x0)
[    1.694776] CPU2: stopping
[    1.700504] CPU: 2 PID: 0 Comm: swapper/2 Not tainted 5.4.31 #0
[    1.703279] Hardware name: Generic DT based system
[    1.709109] [<c030e524>] (unwind_backtrace) from [<c030afa8>] (show_stack+0x10/0x14)
[    1.713972] [<c030afa8>] (show_stack) from [<c0873dd8>] (dump_stack+0x94/0xa8)
[    1.721871] [<c0873dd8>] (dump_stack) from [<c030d7d8>] (handle_IPI+0x164/0x19c)
[    1.728900] [<c030d7d8>] (handle_IPI) from [<c05a8d4c>] (gic_handle_irq+0x8c/0x90)
[    1.736450] [<c05a8d4c>] (gic_handle_irq) from [<c0301a8c>] (__irq_svc+0x6c/0x90)
[    1.743824] Exception stack(0xdf473f60 to 0xdf473fa8)
[    1.751380] 3f60: 00000000 000008b4 dfb551f4 c0313be0 ffffe000 c0c04e68 c0c04ea8 00000004
[    1.756419] 3f80: 00000000 c0b3ede8 c0c0dcf4 00000000 1f015000 df473fb0 c03083e8 c03083ec
[    1.764573] 3fa0: 60000013 ffffffff
[    1.772732] [<c0301a8c>] (__irq_svc) from [<c03083ec>] (arch_cpu_idle+0x38/0x3c)
[    1.776036] [<c03083ec>] (arch_cpu_idle) from [<c03495b0>] (do_idle+0x1e4/0x21c)
[    1.783673] [<c03495b0>] (do_idle) from [<c0349878>] (cpu_startup_entry+0x18/0x1c)
[    1.791053] [<c0349878>] (cpu_startup_entry) from [<803024ac>] (0x803024ac)
[    1.798433] Rebooting in 1 seconds..
```

swg0101 commented:
Thanks - it looks like @chunkeey did add in my changes I was saying earlier.
So in that case, simply doing:
setenv bootargs
saveenv

ldhtrf commented:
Many thanks, the device is now booting correctly :)

There's another minor issue that needs addressing, so far the `rootfs_data` overlay volume does not populate the entire available space on nand if you flash the factory raw UBI image. Simply applying a sysupgrade image resolved this problem.

```
root@OpenWrt:/# ubinfo /dev/ubi0
ubi0
Volumes count:                           3
Logical eraseblock size:                 126976 bytes, 124.0 KiB
Total amount of logical eraseblocks:     932 (118341632 bytes, 112.8 MiB)
Amount of available logical eraseblocks: 397 (50409472 bytes, 48.0 MiB)
Maximum count of volumes                 128
Count of bad physical eraseblocks:       0
Count of reserved physical eraseblocks:  20
Current maximum erase counter value:     1
Minimum input/output unit size:          2048 bytes
Character device major/minor:            246:0
Present volumes:                         0, 1, 2

root@OpenWrt:/# ubinfo /dev/ubi0_2
Volume ID:   2 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        468 LEBs (59424768 bytes, 56.6 MiB)
State:       OK
Name:        rootfs_data
Character device major/minor: 246:3

root@OpenWrt:/# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 2.8M      2.8M         0 100% /rom
tmpfs                   248.2M     64.0K    248.1M   0% /tmp
/dev/ubi0_2              50.6M     40.0K     48.0M   0% /overlay
overlayfs:/overlay       50.6M     40.0K     48.0M   0% /
tmpfs                   512.0K         0    512.0K   0% /dev
```

swg0101 commented:
This makes sense since that's when the partitions are re-initialized. That's generally the first thing I do after doing a raw flash of the device anyways (+ restoring device configs if I had any).

ldhtrf commented:
I have erased mtd12 before flashing the new ubi image, that's why I found it weird that the volume was not auto resized on first boot. The volume does have an autoresize flag though, not sure why it didn't work as intended.
```
$ ubireader_display_info openwrt-ipq40xx-generic-mobipromo_cm520-79f-squashfs-nand-factory.ubi
UBI File
---------------------
	Min I/O: 2048
	LEB Size: 126976
	PEB Size: 131072
	Total Block Count: 45
	Data Block Count: 43
	Layout Block Count: 2
	Internal Volume Block Count: 0
	Unknown Block Count: 0
	First UBI PEB Number: 0

	Image: 1586822102
	---------------------
		Image Sequence Num: 1586822102
		Volume Name:kernel
		Volume Name:rootfs
		Volume Name:rootfs_data
		PEB Range: 0 - 44

		Volume: kernel
		---------------------
			Vol ID: 0
			Name: kernel
			Block Count: 22

			Volume Record
			---------------------
				alignment: 1
				crc: '0x212ca107'
				data_pad: 0
				errors: ''
				flags: 0
				name: 'kernel'
				name_len: 6
				padding: '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
				rec_index: 0
				reserved_pebs: 22
				upd_marker: 0
				vol_type: 'dynamic'


		Volume: rootfs
		---------------------
			Vol ID: 1
			Name: rootfs
			Block Count: 21

			Volume Record
			---------------------
				alignment: 1
				crc: '0x780cb0e7'
				data_pad: 0
				errors: ''
				flags: 0
				name: 'rootfs'
				name_len: 6
				padding: '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
				rec_index: 1
				reserved_pebs: 21
				upd_marker: 0
				vol_type: 'dynamic'


		Volume: rootfs_data
		---------------------
			Vol ID: 2
			Name: rootfs_data
			Block Count: 0

			Volume Record
			---------------------
				alignment: 1
				crc: '0x2913d43c'
				data_pad: 0
				errors: ''
				flags: 'autoresize'
				name: 'rootfs_data'
				name_len: 11
				padding: '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
				rec_index: 2
				reserved_pebs: 9
				upd_marker: 0
				vol_type: 'dynamic'
```

swg0101 commented:
Did you have the set_ubi parameters in uboot when you wrote the image? It's quite possible when you wrote mtd12 it was really only 5x MB because it was set that way in the OEM config...

ldhtrf commented:
Ah, that explains it. Thanks for all the help, I really appreciate it.
